activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -992,6 +992,15 @@ def arel(aliases = nil) # :nodoc:
       @arel ||= build_arel(aliases)
     end
 
+    protected
+      def build_subquery(subquery_alias, select_value) # :nodoc:
+        subquery = except(:optimizer_hints).arel.as(subquery_alias)
```

Comment:
I found this highly surprising. Anyone introducing a `.count` on a relation with distinct or limit/offset that relies on hinting to achieve a good query plan will discover that their hinting is ignored. Worse, they may discover it in production due to performance regression with the unhinted query plan.

Query hints apply to each query block so they aren't safe to shift to other query blocks where they'll have missing table references. For subqueries generated for counting, for example, all optimizer hints (like INDEX, JOIN_PREFIX, etc) reference tables on that relation. Hinting at the top-level query block results in warnings and ignored hints.

I think it'd be clearer to simply apply all optimizer hints to every query block. That'd still result in noisy warnings, but then the hints will be correctly applied to the relation they were declared on.

Reply:
Of the [available optimizer hints](https://dev.mysql.com/doc/refman/8.0/en/optimizer-hints.html) in MySQL 8.0, three have global scope: MAX_EXECUTION_TIME, RESOURCE_GROUP, and SET_VAR.

Reply:
To test with optimizer hints duplicated in the top-level query block and the subquery:
```ruby
module BuildSubqueryWithOptimizerHints
  protected
    def build_subquery(subquery_alias, select_value)
      Arel::SelectManager.new(arel.as(subquery_alias)).project(select_value).tap do |arel|
        arel.optimizer_hints(*optimizer_hints_values) unless optimizer_hints_values.empty?
      end
    end
end

ActiveRecord::QueryMethods.prepend BuildSubqueryWithOptimizerHints
```

This results in warnings, but I think that's our only option. Otherwise Active Record would need to be aware of which hints apply to top-level vs query block.